src/winter_onboarding_2021/fleet_management_service/db/cab.clj

```diff
@@ -1,44 +1,64 @@
 (ns winter-onboarding-2021.fleet-management-service.db.cab
   (:require [winter-onboarding-2021.fleet-management-service.db.core :as db]
             [honey.sql :as sql]
+            [winter-onboarding-2021.fleet-management-service.specs :as specs]
+            [clojure.spec.alpha :as s]
+            [winter-onboarding-2021.fleet-management-service.error :as errors]
+            [winter-onboarding-2021.fleet-management-service.utils :as utils]
             [honey.sql.helpers :as h :refer [select from limit offset order-by update set where]]))
 
-(defn create [cabs]
-  (db/insert! :cabs cabs))
+(defn create [cab]
+  (if (s/valid? ::specs/cabs cab)
+    (db/insert! :cabs cab)
+    errors/validation-failed))
 
 (defn select! [off lim]
-  (db/query! (sql/format (-> (select :id
-                                     :name
-                                     :licence-plate
-                                     :distance-travelled
-                                     :created-at
-                                     :updated-at)
-                             (from :cabs)
-                             (limit lim)
-                             (offset off)
-                             (order-by :created-at)))))
+  (if (s/valid? ::specs/pagination-params {:offset off :limit lim})
+    (db/query! (sql/format (-> (select :id
+                                       :name
+                                       :licence-plate
+                                       :distance-travelled
+                                       :created-at
+                                       :updated-at)
+                               (from :cabs)
+                               (limit lim)
+                               (offset off)
+                               (order-by :created-at))))
+    errors/validation-failed))
 
 (defn get-count []
   (db/query! (sql/format (-> (select [:%count.* :count])
                              (from :cabs)))))
 
 (defn get-by-id [id]
-  {:pre [(uuid? id)]}
-  (db/get-by-id! :cabs id))
+  (if (uuid? id)
+    (db/get-by-id! :cabs id)
+    errors/id-not-uuid))
 
 (defn find-by-keys [key-map]
-  (db/find-by-keys! :cabs key-map))
+  (let [valid-key-map (utils/select-keys-from-spec key-map 
```

Comment:
I understand your point @govindkrjoshi . In this situation I think since we also don't want to pass empty hash-map{} since it will give an `Assertion Error at next.jdbc.sql.builder/by-keys` . So, I was thinking if we could do something like this 
```
(defn find-by-keys [key-map]
  (let [valid-key-map (utils/select-keys-from-spec key-map 
                                                   ::specs/cabs-all-attr)]
    (if (s/valid? ::specs/cabs-all-attr valid-key-map)
      errors/no-valid-keys
      (db/find-by-keys! :cabs valid-key-map))))
```
Instead of using empty? we can use s/valid? which would make it more explicit. wdyt?

Reply:
Since, we discussed, we need to have a check if the `valid-key-map` is not an empty hash-map, because if we allow empty hash map, it will throw an assertion error in the DB. So, if the data has no valid-keys we will send an error saying` no-valid-keys`